### ISSUE TITLE
DeviceClient SetConnectionStatusChangesHandler Not firing when losing Internet

### ISSUE THREAD

**Issue by davehutchy** [Azure/azure-iot-sdk-csharp#350]
- **OS and version used:** ubilinix 4.4.0

- **SDK version used:** 1.6.1

# Description of the issue:
The DeviceClient SetConnectionStatusChangesHandler does not fire when the network goes down. 
However when the network comes back (plug Ethernet back in) the handler fires twice once with a 
status: Disconnected_Retrying
reason: No_Network

and a second time

status: Connected
reason: Connection_Ok 

I was under the impression that the event should fire as soon as network was lost.

# Code sample exhibiting the issue:
```
public async Task RunAsync(Func<TwinCollection, Task<TwinCollection>> twinUpdateHandler, CancellationToken cancellationToken)
{
            
      _client = DeviceClient.Create(_hubOptions.Hostname, new DeviceAuthenticationWithRegistrySymmetricKey(_deviceId, key), TransportType.Amqp_Tcp_Only);

     _client.SetConnectionStatusChangesHandler((status, reason) => _logger.LogWarning($"IoT Hub connection status Changed Status: {status} Reason: {reason}"));

      await _client.OpenAsync();

      await _client.SetDesiredPropertyUpdateCallbackAsync(async (collection, context) =>
                    {
                        var updated = await twinUpdateHandler(collection);
                        await _client.UpdateReportedPropertiesAsync(updated);
                    }
                    , null);

      await ReceiveMessagesAsync(cancellationToken);
}

 private async Task ReceiveMessagesAsync(CancellationToken cancellationToken)
 {
     while (!cancellationToken.IsCancellationRequested)
      {
           var message = await _client.ReceiveAsync(_hubOptions.ReceiveTimeout);
           if (message != null)
           {
                //Do something with received message...
           }
     }
 }
```

# Console log of the issue:
"IoT Hub connection status Changed Status: Connected Reason: Connection_Ok"
------ unplug the internet
------ wait for about 5 minutes
------ plug internet back in
"IoT Hub connection status Changed Status: Disconnected_Retrying Reason: No_Network"
"IoT Hub connection status Changed Status: Connected Reason: Connection_Ok"

**Comment by thesilvereagle** [comment]
Hi there,

I am experiencing a similar issue. We are trying to simulate devices that sends messages to our IoT hub in order to log the exceptions raised when an communcation errors accure as we had several issues with our IoT hub, so that we can send our logs to Microsoft support teams. 
Our main issue is that we SetConnectionStatusChangesHandler is not taken in cosideration (it is supposed to override the one that already exists) when the connection Status changes.

Our main goal is to catch all the exceptions that are raised to have more details in our logs when there is  an interruption.

**Comment by jasmineymlo** [comment]
@davehutchy @thesilvereagle, Thanks for filing the issue.  We will look into this.

**Comment by jasmineymlo** [comment]
@davehutchy @thesilvereagle, just wondering if the devices are only sending telemetry and receiving C2D messages?  Are they subscribed to methods and Twins?  Thanks

**Comment by davehutchy** [comment]
Yes the device does not subscribe to methods however it is subscribed to Twins using  

```
  await _client.SetDesiredPropertyUpdateCallbackAsync(async (collection, context) =>
                    {
                        var updated = await twinUpdateHandler(collection);
                        await _client.UpdateReportedPropertiesAsync(updated);
                    }
                    , null);
```

**Comment by thesilvereagle** [comment]
Thanks @jasmineymlo  for the follow up 
From my side: 
- The devices are only sending telemetry.
- The devices are subscribed to Twins only.

I handled that on my side. We downloaded the SDK and added some code in order to log the exception on a blob so that we can work on the communication errors when any exception is raised.

**Comment by zolvarga** [comment]
@davehutchy 

I have tested with the latest client and here is what is happening:
The client will not fire the ConnectionStatusChange event right away when the network is disconnected. 
But it will fire the event if it tries to send/receive something on the wire. Currently there is a retry mechanism which will trigger the re-connection after 4 min. and that is when the client detects the network problem.
This is the expected behavior.

Could you please test it with the latest client and verify the behavior?

Best Regards,
Zoltan

**Comment by yfakariya** [comment]
@zolvarga @davehutchy 

As far as I tested own small program, these issue ([disconnection notification and restore/disabled notification are fired at once](https://github.com/Azure/azure-iot-sdk-csharp/issues/350#issue-291445472), and [some operations (e.g. DirectMethod) are not working after recovery](https://github.com/Azure/azure-iot-sdk-csharp/issues/350#issuecomment-363047802) ) are not resolve in 1.7.0 SDK.

**Comment by prmathur-microsoft** [comment]
Can you try against 1.20.0 ? We believe we have fixed these kind of issues there.

**Comment by CIPop** [comment]
This should be fixed in 1.19 or 1.20+.

**Comment by az-iot-builder-01** [comment]
@thesilvereagle, @davehutchy, @prmathur-microsoft, @CIPop, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
